Socket buffer metrics

bazsi opened the pull request:
This PR exposes socket buffering related metrics.

kira-syslogng commented:
Build FAILURE

MrAnno commented on modules/afsocket/afsocket-source.c:
```diff
@@ -749,9 +752,7 @@ static void
-
-  if (self->dynamic_window_pool != NULL)
-    _dynamic_window_timer_start(self);
+  _dynamic_window_timer_start(self);
```
I think it is cleaner if `_dynamic_window_timer_start()` actually starts the timer as its name suggests. It is now more of a `_dynamic_window_timer_maybe_start()`, which sounds mystic.

bazsi replied:
I've used the "start" methods similar to how init() is assumed to work, e.g. initialize that portion of the driver, in a similar vein "stop" is parallel do deinit().

The issue is that init is taken and that initializes the low-level parts of the ivykis structs. I didn't change this yet, so keeping this comment open.

MrAnno replied:
Sorry, I don't understand this.
`init()` and `deinit()` does initialization and deinitialization unconditionally.

My comment was a clean-code-related note about a refactor step that made `_dynamic_window_timer_start()` conditional. The function is called timer_start, but after this commit, it sometimes does not start the timer.

MrAnno commented on modules/afsocket/afsocket-source.c:
```diff
@@ -35,6 +35,19 @@
+
+#define SYSLOG_NG_ENABLE_PACKET_STATS 1
+
+static const glong PACKET_STATS_TIMER_MSECS = 1000;
```
Isn't this a very short time if there are a lot of UDP sources or SO_REUSEPORT UDP sources?
Should we make this configurable?

bazsi replied:
I wouldn't want that. maybe 1/sec is a bit too often, but I think we should decide the frequency ourselves. The overhead of the getsockopt in question is not a lot, it's basically a single system call overhead, of which we do 10s of thousands per second, anyway.

MrAnno replied:
Okay. Thank you.

MrAnno commented on modules/afsocket/afsocket-source.c:
```diff
@@ -658,33 +684,88 @@ _dynamic_window_timer_stop(AFSocketSourceDriver *self)
-  if (self->dynamic_window_pool != NULL)
-    _dynamic_window_timer_start(self);
+  _dynamic_window_timer_start(self);
+  _packet_stats_timer_start(self);
```
We'd probably need an immediate `_on_packet_stats_timer_elapsed()` call here if we make the timer interval configurable (so we see max stats at startup/reload quickly).

MrAnno commented on modules/afsocket/afsocket-source.c:
```diff
@@ -1006,11 +1006,20 @@ _register_stats(AFSocketSourceDriver *self)
+          stats_cluster_label("driver_instance", afsocket_sd_format_name(&self->super.super.super)),
+        };
+
+        stats_cluster_single_key_set(&sc_key, "source_connections_total", labels, G_N_ELEMENTS(labels));
```
Sorry, I was wrong: `total` is for accumulating counters.
I think this should be: `input_connections`.

bazsi replied:
We are calling similar counters for named log statements as "ingress" and "egress" so we might want to review that later. But I have now renamed this counter to input_connections as you requested.

bazsi replied:
The log-path specific metric is called log_path_ingress and log_path_egress, note that there's no "total" suffix here.

maybe we should call the source related counters source_ingress? Also, we should either use total in both cases or in neither ones.

What do you think?

MrAnno replied:
`total` is needed for accumulating counters, so I think we should fix the metric pipe, but we should not add `total` here as `input_connections` is not an accumulation.

I'd avoid using the word `source`, as we started using `input` instead, but we can change that if you don't like it.
`source_ingress` sounds redundant to me, but I might misunderstood the meaning.

bazsi commented on modules/afsocket/afsocket-source.c:
```diff
@@ -658,17 +671,77 @@ _dynamic_window_timer_stop(AFSocketSourceDriver *self)
+static void
+_packet_stats_timer_start(AFSocketSourceDriver *self)
+{
+  if (!self->stat_socket_dropped_packets)
```
you are right, I was moving that _register_stats() call back and forth while refactoring and I didn't test the end result.

bazsi replied:
fixed.

MrAnno replied:
Thanks.

bazsi commented:
Fixed most of the outstanding issues (those that I resolved above). After the changes I have now reworked the branch moving the refactoring steps first and then the functional changes. This makes the branch more difficult to re-review, but should make the branch easier to understand, losing history on some of the issues you pointed out.

I've produced a diff between the state you reviewed and the current one, which I am attaching here.

bazsi commented:
And the patch:

```
diff --git a/configure.ac b/configure.ac
index f2fd413e4..c6b281cb3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -787,6 +787,22 @@ if test "x$blb_cv_c_modern_utmp" = "xyes"; then
 	AC_DEFINE(HAVE_MODERN_UTMP, 1, [have modern utmp/utmpx format])
 fi
 
+old_CPPFLAGS=$CPPFLAGS
+CPPFLAGS="-D_GNU_SOURCE -D_DEFAULT_SOURCE"
+AC_CACHE_CHECK(for SO_MEMINFO, blb_cv_c_so_meminfo,
+  [AC_EGREP_CPP(SO_MEMINFO,
+[
+#include <sys/socket.h>
+SO_MEMINFO
+],
+  blb_cv_c_so_meminfo=no, blb_cv_c_so_meminfo=yes)])
+CPPFLAGS=$old_CPPFLAGS
+
+if test "x$blb_cv_c_so_meminfo" = "xyes"; then
+	AC_DEFINE(HAVE_SO_MEMINFO, 1, [have SO_MEMINFO sockopt])
+	AC_DEFINE(ENABLE_AFSOCKET_MEMINFO_METRICS, 1, [enable SO_MEMINFO based metrics in afsocket])
+fi
+
 dnl ***************************************************************************
 dnl Checks for libraries
 AC_CHECK_LIB(m, round, BASE_LIBS="$BASE_LIBS -lm")
diff --git a/modules/afsocket/afsocket-source.c b/modules/afsocket/afsocket-source.c
index 428221724..12634fef7 100644
--- a/modules/afsocket/afsocket-source.c
+++ b/modules/afsocket/afsocket-source.c
@@ -35,17 +35,12 @@
 #include <sys/types.h>
 #include <sys/socket.h>
 
-#ifdef SYSLOG_NG_HAVE_LINUX_SOCK_DIAG_H
+#if SYSLOG_NG_HAVE_LINUX_SOCK_DIAG_H
 #include <linux/sock_diag.h>
-
-#ifdef SO_MEMINFO
-
-#define SYSLOG_NG_ENABLE_PACKET_STATS 1
-
-static const glong PACKET_STATS_TIMER_MSECS = 1000;
-
 #endif
 
+#if SYSLOG_NG_ENABLE_AFSOCKET_MEMINFO_METRICS
+static const glong PACKET_STATS_TIMER_MSECS = 1000;
 #endif
 
 #if SYSLOG_NG_ENABLE_TCP_WRAPPER
@@ -685,8 +680,7 @@ _dynamic_window_timer_stop(AFSocketSourceDriver *self)
     iv_timer_unregister(&self->dynamic_window_timer);
 }
 
-
-#ifdef SYSLOG_NG_ENABLE_PACKET_STATS
+#if SYSLOG_NG_ENABLE_AFSOCKET_MEMINFO_METRICS
 
 static void
 _packet_stats_timer_start(AFSocketSourceDriver *self)
@@ -736,11 +730,43 @@ _packet_stats_timer_stop(AFSocketSourceDriver *self)
     iv_timer_unregister(&self->packet_stats_timer);
 }
 
+static void
+_register_packet_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels, gsize labels_len)
+{
+  StatsClusterKey sc_key;
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_dropped_packets_total", labels, labels_len);
+  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_dropped_packets);
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_max_bytes", labels, labels_len);
+  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_max);
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_used_bytes", labels, labels_len);
+  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_used);
+}
+
+static void
+_unregister_packet_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels, gsize labels_len)
+{
+  StatsClusterKey sc_key;
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_dropped_packets_total", labels, labels_len);
+  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_dropped_packets);
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_max_bytes", labels, labels_len);
+  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_max);
+
+  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_used_bytes", labels, labels_len);
+  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_used);
+}
+
 #else
 
 #define _packet_stats_timer_init(s)
 #define _packet_stats_timer_start(s)
 #define _packet_stats_timer_stop(s)
+#deifne _register_packet_stats(s, l, ll)
+#deifne _unregister_packet_stats(s, l, ll)
 
 #endif
 
@@ -748,8 +774,8 @@ static void
 _init_watches(AFSocketSourceDriver *self)
 {
   _dynamic_window_timer_init(self);
-  _packet_stats_timer_init(self);
   _listen_fd_init(self);
+  _packet_stats_timer_init(self);
 }
 
 static void
@@ -866,7 +892,33 @@ afsocket_sd_setup_transport(AFSocketSourceDriver *self)
   return TRUE;
 }
 
-static gboolean
+static void
+afsocket_sd_save_connections(AFSocketSourceDriver *self)
+{
+  GlobalConfig *cfg = log_pipe_get_config(&self->super.super.super);
+
+  if (!self->connections_kept_alive_across_reloads || !cfg->persist)
+    {
+      afsocket_sd_kill_connection_list(self->connections);
+    }
+  else
+    {
+      GList *p;
+
+      /* for SOCK_STREAM source drivers this is a list, for
+       * SOCK_DGRAM this is a single connection */
+
+      for (p = self->connections; p; p = p->next)
+        {
+          log_pipe_deinit((LogPipe *) p->data);
+        }
+      cfg_persist_config_add(cfg, afsocket_sd_format_connections_name(self), self->connections,
+                             (GDestroyNotify)afsocket_sd_kill_connection_list, FALSE);
+    }
+  self->connections = NULL;
+}
+
+static void
 afsocket_sd_restore_kept_alive_connections(AFSocketSourceDriver *self)
 {
   GlobalConfig *cfg = log_pipe_get_config(&self->super.super.super);
@@ -894,7 +946,6 @@ afsocket_sd_restore_kept_alive_connections(AFSocketSourceDriver *self)
             }
         }
     }
-  return TRUE;
 }
 
 static gboolean
@@ -992,31 +1043,6 @@ afsocket_sd_close_fd(gpointer value)
   close(fd);
 }
 
-static void
-afsocket_sd_save_connections(AFSocketSourceDriver *self)
-{
-  GlobalConfig *cfg = log_pipe_get_config(&self->super.super.super);
-
-  if (!self->connections_kept_alive_across_reloads || !cfg->persist)
-    {
-      afsocket_sd_kill_connection_list(self->connections);
-    }
-  else
-    {
-      GList *p;
-
-      /* for SOCK_STREAM source drivers this is a list, for
-       * SOCK_DGRAM this is a single connection */
-
-      for (p = self->connections; p; p = p->next)
-        {
-          log_pipe_deinit((LogPipe *) p->data);
-        }
-      cfg_persist_config_add(cfg, afsocket_sd_format_connections_name(self), self->connections,
-                             (GDestroyNotify)afsocket_sd_kill_connection_list, FALSE);
-    }
-  self->connections = NULL;
-}
 
 static void
 afsocket_sd_save_listener(AFSocketSourceDriver *self)
@@ -1090,6 +1116,16 @@ afsocket_sd_create_dynamic_window_pool(AFSocketSourceDriver *self)
     }
 }
 
+static void
+afsocket_sd_drop_dynamic_window_pool(AFSocketSourceDriver *self)
+{
+  if (self->dynamic_window_pool)
+    {
+      dynamic_window_pool_unref(self->dynamic_window_pool);
+      self->dynamic_window_pool = NULL;
+    }
+}
+
 gboolean
 afsocket_sd_setup_addresses_method(AFSocketSourceDriver *self)
 {
@@ -1101,7 +1137,7 @@ _register_stream_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels, gs
 {
   StatsClusterKey sc_key;
 
-  stats_cluster_single_key_set(&sc_key, "source_connections_total", labels, labels_len);
+  stats_cluster_single_key_set(&sc_key, "input_connections", labels, labels_len);
 
   stats_cluster_single_key_add_legacy_alias_with_name(&sc_key,
                                                       self->transport_mapper->stats_source | SCS_SOURCE,
@@ -1117,7 +1153,7 @@ _unregister_stream_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels,
 {
   StatsClusterKey sc_key;
 
-  stats_cluster_single_key_set(&sc_key, "source_connections_total", labels, labels_len);
+  stats_cluster_single_key_set(&sc_key, "input_connections", labels, labels_len);
 
   stats_cluster_single_key_add_legacy_alias_with_name(&sc_key,
                                                       self->transport_mapper->stats_source | SCS_SOURCE,
@@ -1130,36 +1166,13 @@ _unregister_stream_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels,
 static void
 _register_dgram_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels, gsize labels_len)
 {
-#ifdef SYSLOG_NG_ENABLE_PACKET_STATS
-  StatsClusterKey sc_key;
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_dropped_packets_total", labels, labels_len);
-  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_dropped_packets);
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_max_bytes", labels, labels_len);
-  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_max);
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_used_bytes", labels, labels_len);
-  stats_register_counter(1, &sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_used);
-#endif
+  _register_packet_stats(self, labels, labels_len);
 }
 
 static void
 _unregister_dgram_stats(AFSocketSourceDriver *self, StatsClusterLabel *labels, gsize labels_len)
 {
-#ifdef SYSLOG_NG_ENABLE_PACKET_STATS
-  StatsClusterKey sc_key;
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_dropped_packets_total", labels, labels_len);
-  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_dropped_packets);
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_max_bytes", labels, labels_len);
-  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_max);
-
-  stats_cluster_single_key_set(&sc_key, "socket_receive_buffer_used_bytes", labels, labels_len);
-  stats_unregister_counter(&sc_key, SC_TYPE_SINGLE_VALUE, &self->stat_socket_receive_buffer_used);
-#endif
-
+  _unregister_packet_stats(self, labels, labels_len);
 }
 
 static void
@@ -1208,12 +1221,18 @@ afsocket_sd_init_method(LogPipe *s)
   if (!afsocket_sd_restore_dynamic_window_pool(self))
     afsocket_sd_create_dynamic_window_pool(self);
 
-  if (!(afsocket_sd_restore_kept_alive_connections(self) &&
-        afsocket_sd_open_listener(self)))
-    return FALSE;
+  afsocket_sd_restore_kept_alive_connections(self);
 
   _register_stats(self);
 
+  if (!afsocket_sd_open_listener(self))
+    {
+      /* returning FALSE, so deinit is not called */
+      _unregister_stats(self);
+      afsocket_sd_drop_dynamic_window_pool(self);
+      return FALSE;
+    }
+
   return TRUE;
 }
 
diff --git a/syslog-ng/main.c b/syslog-ng/main.c
index 0d924a01b..6bc1ee869 100644
--- a/syslog-ng/main.c
+++ b/syslog-ng/main.c
@@ -306,18 +306,16 @@ main(int argc, char *argv[])
   main_loop_init(main_loop, &main_loop_options);
   rc = main_loop_read_and_init_config(main_loop);
 
-  if (rc)
+  if (rc || exit_before_main_loop_run)
     {
+      main_loop_deinit(main_loop);
+      app_shutdown();
+      reloc_deinit();
       g_process_startup_failed(rc, TRUE);
       return rc;
     }
-  else
-    {
-      if (exit_before_main_loop_run)
-        g_process_startup_failed(0, TRUE);
-      else
-        g_process_startup_ok();
-    }
+
+  g_process_startup_ok();
 
   /* we are running as a non-root user from this point */

```

bazsi commented:
I still need to test this stuff, as I moved lot of patched around and I only validated the branch with my eyes.

kira-syslogng commented:
Build FAILURE

bazsi commented:
this branch has gone through so much back and forth, so it is definitely easier to re-review the whole thing.

It includes a number of refactoring steps and then exposes two prometheus style metrics:
  * the number of open connections to an afsocket based source (e.g. TCP using transports)
  * three metrics on dropped packets, maximum and used buffer space (e.g. UDP using transports)

The new metric names now match the ones we've discussed IRL.

UDP metrics:
```
# syslog-ng-ctl  stats prometheus | grep socket 
syslogng_socket_receive_buffer_used_bytes{id="#anon-source0#3",direction="input",driver_instance="afsocket_sd.udp4"} 0
syslogng_socket_receive_buffer_max_bytes{id="#anon-source0#0",direction="input",driver_instance="afsocket_sd.udp1"} 268435456
syslogng_socket_receive_dropped_packets_total{id="#anon-source0#0",direction="input",driver_instance="afsocket_sd.udp1"} 0
syslogng_socket_receive_buffer_max_bytes{id="#anon-source0#1",direction="input",driver_instance="afsocket_sd.udp2"} 212992
syslogng_socket_receive_dropped_packets_total{id="#anon-source0#1",direction="input",driver_instance="afsocket_sd.udp2"} 0
syslogng_socket_receive_buffer_max_bytes{id="#anon-source0#2",direction="input",driver_instance="afsocket_sd.udp3"} 212992
syslogng_socket_receive_dropped_packets_total{id="#anon-source0#2",direction="input",driver_instance="afsocket_sd.udp3"} 0
syslogng_socket_receive_buffer_used_bytes{id="#anon-source0#0",direction="input",driver_instance="afsocket_sd.udp1"} 0
syslogng_socket_receive_buffer_max_bytes{id="#anon-source0#3",direction="input",driver_instance="afsocket_sd.udp4"} 212992
syslogng_socket_receive_dropped_packets_total{id="#anon-source0#3",direction="input",driver_instance="afsocket_sd.udp4"} 619173
syslogng_socket_receive_buffer_used_bytes{id="#anon-source0#1",direction="input",driver_instance="afsocket_sd.udp2"} 0
syslogng_socket_receive_buffer_used_bytes{id="#anon-source0#2",direction="input",driver_instance="afsocket_sd.udp3"} 0
```

TCP metrics:
```
# syslog-ng-ctl  stats prometheus | grep conne
syslogng_socket_connections{id="#anon-source0#0",direction="input",driver_instance="afsocket_sd.(stream,AF_INET(0.0.0.0:2000))"} 1
```

as you can see the "driver_instance" label still needs to get exploded into labels but that should be a separate PR.

bazsi commented on modules/afsocket/afsocket-source.c:
```diff
@@ -1220,9 +1223,7 @@ afsocket_sd_deinit_method(LogPipe *s)
-
-  if (self->dynamic_window_pool)
-    afsocket_sd_save_dynamic_window_pool(self);
+  afsocket_sd_save_dynamic_window_pool(self);
```
I have tried to address this discussion in the latest incarnation of the patches.

Low level functions associated with the listener, dynamic-window and packet-stats don't use the afsocket_sd_ prefix, while higher level functions (those that work on integrating the lower level support code into the larger afsocket), do use the prefix.

This makes it a bit easier to navigate the code. I am not saying it's perfect from a clean code perspective, but I think it did improve a lot, trying to concentrate related code to the same spot and trying to use consistent naming/prefixes.

Unless I broke something I'd really want to integrate this in this form, I spent too much time in juggling with the patches and feel I spent too much time on it already.

MrAnno reviewed (approved):
Perfect, thank you.
I've added a NEWS entry.

MrAnno commented:
@kira-syslogng do stresstest

MrAnno commented:
Waiting for Kira to reappear.

@HofiOne @folti

HofiOne commented:
> Waiting for Kira to reappear.
> 
> @HofiOne @folti

the issue is solved AFAIK

kira-syslogng commented:
Kira-stress-test: Build FAILURE

kira-syslogng commented:
Build FAILURE

MrAnno commented:
@kira-syslogng retest this please

MrAnno commented:
@kira-syslogng do stresstest

kira-syslogng commented:
Kira-stress-test: Build SUCCESS
